File: src/InfoProviders/InfoLookupProvider.py
import json
import os
from enum import Enum

import pycountry as pc
import yfinance as yf

rootOfProject = os.getcwd()
cacheDir = os.path.join(rootOfProject, "cache", "yfinance")
yf.set_tz_cache_location(cacheDir)


class TreatyType(str, Enum):
    TaxRelief = "TAX_RELIEF"


class Country:
    name: str
    shortCode2: str
    treaties: dict[TreatyType, str]
    # https://www.gov.si/drzavni-organi/ministrstva/ministrstvo-za-finance/o-ministrstvu/direktorat-za-sistem-davcnih-carinskih-in-drugih-javnih-prihodkov/seznam-veljavnih-konvencij-o-izogibanju-dvojnega-obdavcevanja-dohodka-in-premozenja/
    # https://www.fu.gov.si/davki_in_druge_dajatve/podrocja/mednarodno_obdavcenje/#c78

    def __init__(self, name: str, shortCode2: str, treaties: dict[TreatyType, str]):
        self.name = name
        self.shortCode2 = shortCode2
        self.treaties = treaties


class CountryLookupProvider:
    problematicCountryMappings: dict[str, str]

    def __init__(self):
        definitionsPath = os.path.join(
            os.getcwd(), "src", "InfoProviders", "internationalTreaties.json"
        )
        problematicCountryMappingsFile = os.path.join(
            os.getcwd(), "src", "InfoProviders", "specialCountryMappings.json"
        )

        with open(definitionsPath, "r") as read_file:
            self.definitions = json.load(read_file)

        with open(problematicCountryMappingsFile, "r") as read_file:
            self.problematicCountryMappings = json.load(read_file)

    def getCountry(self, country: str) -> Country:
        if self.problematicCountryMappings.get(country) is None:
            countryDefinition = pc.countries.get(name=country)
            shortCode = countryDefinition.alpha_2
        else:
            shortCode = self.problematicCountryMappings[country]

        definedTreaties = dict(self.definitions["treaties"]["taxRelief"])
        thisCountryTreaties: dict[TreatyType, str] = {}

        if definedTreaties.get(shortCode) is not None:
            thisCountryTreaties[TreatyType.TaxRelief] = str(definedTreaties[shortCode])

        return Country(country, shortCode, thisCountryTreaties)


countryLookupInstance = CountryLookupProvider()


class CompanyLocationInfo:
    Country: str
    Address1: str
    Address2: str | None
    ZipCode: str
    City: str
    State: str | None

    ShortCodeCountry2: str

    def __init__(
        self,
        country: str,
        address1: str,
        address2: str | None,
        zipCode: str,
        city: str,
        state: str | None,
    ):
        self.Country = country
        self.Address1 = address1
        self.Address2 = address2
        self.ZipCode = zipCode
        self.City = city
        self.State = state

        countryDefinition = countryLookupInstance.getCountry(country)

        self.ShortCodeCountry2 = countryDefinition.shortCode2

    def formatAsUnternationalAddress(self) -> str:
        startingAddress = "{},".format(self.Address1)
        if self.Address2 is not None:
            startingAddress = "{}, {},".format(self.Address1, self.Address2)

        address = "{} {},".format(startingAddress, self.City)

        if self.State is not None:
            address = "{} {}".format(address, self.State)

        address = "{} {}, {}".format(address, self.ZipCode, self.Country)

        return address


class CompanyInfo:
    Location: CompanyLocationInfo
    ShortName: str
    LongName: str

    def __init__(self, shortName: str, longName: str, location: CompanyLocationInfo):
        self.ShortName = shortName
        self.LongName = longName
        self.Location = location


class CompanyLookupProvider:
    mappings: dict[str, yf.Ticker] = dict()
    isinToTickerLookup: dict[str, str]
    isinToCompanyLookup: dict[str, dict[str, str]]

    def __init__(self):
        definitionsPath = os.path.join(
            os.getcwd(), "src", "InfoProviders", "missingISINLookup.json"
        )

        with open(definitionsPath, "r") as read_file:
            self.isinToTickerLookup = dict(json.load(read_file))

        definitionsPath = os.path.join(
            os.getcwd(), "src", "InfoProviders", "missingCompaniesLookup.json"
        )

        with open(definitionsPath, "r") as read_file:
            self.isinToCompanyLookup = dict(json.load(read_file))

    def getCompanyInfo(self, isin: str) -> CompanyInfo:
        try:
            yfinanceTicker = self.getInfoByISIN(isin)
            companyInfo = yfinanceTicker.info
            if companyInfo.values().__len__() == 0:
                raise ValueError("Empty Company Data for ISIN ({})".format(isin))
        except Exception:
            print(
                "Failed online lookup of ISIN ({}), falling back to hardcoded company definitions".format(
                    isin
                )
            )

            if self.isinToCompanyLookup.get(isin) is None:
                raise ValueError("Failed lookup of ISIN ({})".format(isin))

            companyInfo = self.isinToCompanyLookup[isin]

        address1 = companyInfo.get("address1")
        address2 = companyInfo.get("address2")
        city = companyInfo.get("city")
        state = companyInfo.get("state")
        zipCode = companyInfo.get("zip")
        country = companyInfo.get("country")
        locationInfo = CompanyLocationInfo(
            country, address1, address2, zipCode, city, state
        )

        shortName = companyInfo.get("shortName")
        longName = companyInfo.get("longName")
        company = CompanyInfo(shortName, longName, locationInfo)
        return company

    def getInfoByTicker(self, ticker: str) -> yf.Ticker:
        if self.mappings.get(ticker) is not None:
            return self.mappings[ticker]

        lookUp = yf.Ticker(ticker)
        self.mappings[ticker] = lookUp
        return lookUp

    def getInfoByISIN(self, isin: str) -> yf.Ticker:
        try:
            return self.getInfoByTicker(isin)
        except Exception as e:
            print(
                "Failed lookup using ISIN ({}), trying again with ISIN mapping if exists".format(
                    isin
                )
            )
            print("Cause: {}".format(e))

        if self.isinToTickerLookup.get(isin) is None:
            raise ValueError("ISIN {} has no mapping to fall back on".format(isin))

        return self.getInfoByTicker(self.isinToTickerLookup[isin])

File: lib/InfoProviders/InfoLookupProvider.ts
import internationalTreaties from "./internationalTreaties.json" with { type: "json" };
import specialCountryMappings from "./specialCountryMappings.json" with { type: "json" };
import missingISINLookup from "./missingISINLookup.json" with { type: "json" };
import missingCompaniesLookup from "./missingCompaniesLookup.json" with { type: "json" };

export enum TreatyType {
	TaxRelief = "TAX_RELIEF",
}

export type Country = {
	name: string;
	shortCode2: string;
	treaties: Map<TreatyType, string>;
};

export type CompanyLocationInfo = {
	country: string;
	address1: string;
	address2: string | null;
	zipCode: string;
	city: string;
	state: string | null;
	shortCodeCountry2: string;
};

export type CompanyInfo = {
	shortName: string;
	longName: string;
	location: CompanyLocationInfo;
};

const problematicCountryMappings = specialCountryMappings as Record<string, string>;
const taxReliefTreaties = (internationalTreaties as { treaties: { taxRelief: Record<string, string> } }).treaties
	.taxRelief;

// ISO 3166-1 country name → alpha-2 code (subset of common countries; extend as needed)
const ISO_COUNTRY_NAME_TO_ALPHA2: Record<string, string> = {
	"Afghanistan": "AF",
	"Albania": "AL",
	"Algeria": "DZ",
	"Andorra": "AD",
	"Angola": "AO",
	"Argentina": "AR",
	"Armenia": "AM",
	"Australia": "AU",
	"Austria": "AT",
	"Azerbaijan": "AZ",
	"Bahrain": "BH",
	"Bangladesh": "BD",
	"Belarus": "BY",
	"Belgium": "BE",
	"Benin": "BJ",
	"Bolivia": "BO",
	"Bosnia and Herzegovina": "BA",
	"Brazil": "BR",
	"Bulgaria": "BG",
	"Cambodia": "KH",
	"Cameroon": "CM",
	"Canada": "CA",
	"Chile": "CL",
	"China": "CN",
	"Colombia": "CO",
	"Croatia": "HR",
	"Cyprus": "CY",
	"Czech Republic": "CZ",
	"Czechia": "CZ",
	"Denmark": "DK",
	"Ecuador": "EC",
	"Egypt": "EG",
	"Estonia": "EE",
	"Ethiopia": "ET",
	"Finland": "FI",
	"France": "FR",
	"Georgia": "GE",
	"Germany": "DE",
	"Ghana": "GN",
	"Greece": "GR",
	"Guatemala": "GT",
	"Hong Kong": "HK",
	"Hungary": "HU",
	"Iceland": "IS",
	"India": "IN",
	"Indonesia": "ID",
	"Iran": "IR",
	"Iraq": "IQ",
	"Ireland": "IE",
	"Israel": "IL",
	"Italy": "IT",
	"Japan": "JP",
	"Jordan": "JO",
	"Kazakhstan": "KZ",
	"Kenya": "KE",
	"Kuwait": "KW",
	"Latvia": "LV",
	"Lebanon": "LB",
	"Lithuania": "LT",
	"Luxembourg": "LU",
	"Malaysia": "MY",
	"Malta": "MT",
	"Mexico": "MX",
	"Moldova": "MD",
	"Morocco": "MA",
	"Netherlands": "NL",
	"New Zealand": "NZ",
	"Nigeria": "NG",
	"North Macedonia": "MK",
	"Norway": "NO",
	"Oman": "OM",
	"Pakistan": "PK",
	"Panama": "PA",
	"Peru": "PE",
	"Philippines": "PH",
	"Poland": "PL",
	"Portugal": "PT",
	"Qatar": "QA",
	"Romania": "RO",
	"Russia": "RU",
	"Russian Federation": "RU",
	"Saudi Arabia": "SA",
	"Serbia": "RS",
	"Singapore": "SG",
	"Slovakia": "SK",
	"Slovenia": "SI",
	"South Korea": "KR",
	"Republic of Korea": "KR",
	"Spain": "ES",
	"Sri Lanka": "LK",
	"Sweden": "SE",
	"Switzerland": "CH",
	"Taiwan": "TW",
	"Thailand": "TH",
	"Turkey": "TR",
	"Turkiye": "TR",
	"Ukraine": "UA",
	"United Arab Emirates": "AE",
	"United Kingdom": "GB",
	"United States": "US",
	"Uruguay": "UY",
	"Venezuela": "VE",
	"Vietnam": "VN",
	"Zambia": "ZM",
	"Zimbabwe": "ZW",
};

export class CountryLookupProvider {
	getCountry(country: string): Country {
		let shortCode: string;
		if (problematicCountryMappings[country] !== undefined) {
			shortCode = problematicCountryMappings[country];
		} else if (ISO_COUNTRY_NAME_TO_ALPHA2[country] !== undefined) {
			shortCode = ISO_COUNTRY_NAME_TO_ALPHA2[country];
		} else {
			throw new Error(`Unknown country: ${country}. Add it to specialCountryMappings.json`);
		}

		const treaties = new Map<TreatyType, string>();
		if (taxReliefTreaties[shortCode] !== undefined) {
			treaties.set(TreatyType.TaxRelief, taxReliefTreaties[shortCode]);
		}

		return { name: country, shortCode2: shortCode, treaties };
	}
}

const isinToTickerLookup = missingISINLookup as Record<string, string>;
const isinToCompanyLookup = missingCompaniesLookup as Record<
	string,
	{
		shortName: string;
		longName: string;
		address1: string;
		address2?: string;
		city: string;
		zip: string;
		country: string;
		state?: string;
	}
>;

const countryLookupInstance = new CountryLookupProvider();

export class CompanyLookupProvider {
	getCompanyInfo(isin: string): CompanyInfo {
		const companyData = isinToCompanyLookup[isin] ?? isinToCompanyLookup[isinToTickerLookup[isin] ?? ""];

		if (companyData === undefined) {
			throw new Error(`Failed lookup of ISIN (${isin})`);
		}

		const countryDef = countryLookupInstance.getCountry(companyData.country);

		const location: CompanyLocationInfo = {
			country: companyData.country,
			address1: companyData.address1,
			address2: companyData.address2 ?? null,
			zipCode: companyData.zip,
			city: companyData.city,
			state: companyData.state ?? null,
			shortCodeCountry2: countryDef.shortCode2,
		};

		return {
			shortName: companyData.shortName,
			longName: companyData.longName,
			location,
		};
	}
}
